packages/front-end/pages/saved-groups.tsx

```diff
@@ -78,6 +116,20 @@ export default function SavedGroupsPage() {
               </thead>
               <tbody>
                 {savedGroups.map((s) => {
+                  const featuresUsingSavedGroups: string[] = [];
+
+                  features.forEach((feature) => {
+                    for (const env in feature.environmentSettings) {
+                      feature.environmentSettings[env].rules.forEach((rule) => {
+                        if (
+                          rule.condition.includes(s.id) &&
+                          !featuresUsingSavedGroups.includes(feature.id)
+                        ) {
+                          featuresUsingSavedGroups.push(feature.id);
+                        }
+                      });
+                    }
+                  });
```

Comment:
This would probably read better as a memoized function outside of the template, e.g.:

```ts
const featuresUsingSavedGroups = useMemo(() => {
  const featuresUsingSavedGroups: string[] = [];

  features.forEach((feature) => {
    for (const env in feature.environmentSettings) {
      feature.environmentSettings[env].rules.forEach((rule) => {
        if (
          rule.condition.includes(s.id) &&
          !featuresUsingSavedGroups.includes(feature.id)
        ) {
          featuresUsingSavedGroups.push(feature.id);
        }
      });
    }
  });

  return featuresUsingSavedGroups
}, [savedGroups])
```

That will allow us to move logic out of the template which will improve readability. It also means the `forEach` will be memoized so it doesn't get executed on every re-render, only when `savedGroups` changes.